### PR TITLE
FEAT: Add simple evaluation functions

### PR DESCRIPTION
**Pull request by AubaneNourry**
Simple evaluation functions for 20x20 Gomoku boards.
Will probably change to consider other indicators like:
- Mobility Score (Positional Advantage)
- Weighted Threat Assessment
In other set of functions.

See : 
- #10 for more informations.

### CODE REVIEW

File: evaluation.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+#avg 0.00025 sec
+def evaluate_pattern_based(board: list, player: int = 1) -> int:
+    score = 0
+    opponent = -player
+
+    rows = board
+    cols = [[board[r][c] for r in range(len(board))] for c in range(len(board[0]))]
+    diags = get_diagonals(board)
```
Comment (Peralban) [inline evaluation.py:29]: Types

File: evaluation.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        score -= score_line(line, opponent) // 2  # Prefer winning instead of avoiding loss
+
+    return score
+
+
+def score_line(line: list, player: int) -> int:
+    """Count consecutive pieces in a line and assign scores based on pattern"""
+    count = 0
+    open_ends = 0
+    line_score = 0
```
Comment (Peralban) [inline evaluation.py:42]: Same

File: evaluation.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+    i = 0
```
Comment (Peralban) [inline evaluation.py:44]: Same

File: evaluation.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            if count > 0:
+                line_score += calculate_score(count, open_ends)
+                open_ends += 1
+                count = 0
+            else:
+                open_ends = 1
+        else:
+            if count > 0:
+                line_score += calculate_score(count, open_ends)
+            count, open_ends = 0, 0
```
Comment (Peralban) [inline evaluation.py:58]: elif
else ?

File: evaluation.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+def calculate_score(count: int, open_ends: int) -> int:
+    """Use pattern scores dictionary to get score based on piece count and open ends"""
+    return PATTERN_SCORES.get((count, open_ends), 0)
+
+
+def get_diagonals(board: list) -> list:
+    """Extract all diagonals (both directions) from the board"""
+    n, m = len(board), len(board[0])
+    diags = []
```
Comment (Peralban) [inline evaluation.py:75]: Types

File: evaluation.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+    for d in range(n + m - 1):
+        diag1, diag2 = [], []
```
Comment (Peralban) [inline evaluation.py:78]: Types

File: evaluation.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        for i in range(max(0, d - m + 1), min(n, d + 1)):
+            diag1.append(board[i][d - i])
+            diag2.append(board[n - i - 1][d - i])
+        diags.append(diag1)
+        diags.append(diag2)
+    return diags
+
+#avg 0.0001 sec
+def evaluate_line_of_sight(board: list, player: int = 1) -> int:
+    score = 0
```
Comment (Peralban) [inline evaluation.py:88]: Same

File: evaluation.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    for col in zip(*board):
+        score += score_line(col, player)
+    for diag in get_diagonals(board):
+        score += score_line(diag, player)
+    return score
+
+#avg 0.00020 sec
+def evaluate_line_of_sight_with_opponent(board: list, player: int = 1) -> int:
+    score = 0
+    opponent = -player
```
Comment (Peralban) [inline evaluation.py:100]: Same

**Review by Peralban — CHANGES_REQUESTED** [review]

File: evaluation.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    (4, 0): 1000,    # Closed Four (4 pieces in a row but blocked on both ends)
+    (3, 2): 1000,    # Open Three (3 pieces in a row with two open ends)
+    (3, 1): 100,     # Closed Three (3 pieces in a row with one open end)
+    (2, 2): 100,     # Open Two (2 pieces in a row with two open ends)
+    (2, 1): 10,      # Closed Two (2 pieces in a row with one open end)
+    (1, 2): 1        # Single piece with two open ends
+}
+
+#avg 0.00025 sec
+def evaluate_pattern_based(board: list, player: int = 1) -> int:
```
Comment (Thomaltarix) [inline evaluation.py:23]: You can a comment the function 😉

File: evaluation.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    for row in board:
+        score += score_line(row, player)
+    for col in zip(*board):
+        score += score_line(col, player)
+    for diag in get_diagonals(board):
+        score += score_line(diag, player)
+    return score
+
+#avg 0.00020 sec
+def evaluate_line_of_sight_with_opponent(board: list, player: int = 1) -> int:
```
Comment (Thomaltarix) [inline evaluation.py:98]: You can comment this one too

**Review by Thomaltarix — CHANGES_REQUESTED** [review]
Seems fair

File: evaluation.py (hunk 10 above)
Comment (AubaneNourry) [reply evaluation.py:98]: why? these two are the most accurate since line of sight does not count the opponent pieces in the returned value?

Comment (Thomaltarix) [reply evaluation.py]: The main reason is to know what are the parameters that you need to send. For example, the board is a list, so it's obvious. But for the player, without reading the code, how do you know what you're supposed to send as parameter for the player ? (And what means the 1 that is by default ?)

**Review by Peralban — DISMISSED** [review]

**Review by Thomaltarix — DISMISSED** [review]
I'll let you just add the documentation, but that's great

**Review by Thomaltarix — APPROVED** [review]

**Pull request merged**